lib/spack/spack/util/download.py

```diff
@@ -0,0 +1,345 @@
+    url: str
+
+    #: Header of the final response.
+    headers: HTTPMessage
+
+    @abc.abstractmethod
+    def read(self, size: int) -> bytes:
+        """Reads and returns a chunk of the final response body."""
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def geturl(self) -> str:
+        """Returns the effective URL of the request"""
+        raise NotImplementedError
+
+
+class UrllibStream(UrlStream):
+    def __init__(self, *, url: str, response):
+        # We cannot return a raw response because we need the original url in some place
+        # and HTTPResponse.url is the same as geturl()
+        self.url = url
+        self.headers = response.headers
+        self._response = response
+
+    def read(self, size: int) -> bytes:
+        return self._response.read(size)
+
+    def geturl(self) -> str:
+        return self._response.geturl()
+
+
+@contextlib.contextmanager
+def urllib_stream(url: str, timeout: int) -> Generator[UrlStream, Any, None]:
+    request = urllib.request.Request(url, headers={"User-Agent": SPACK_USER_AGENT})
+    with urlopen(request, timeout=timeout) as response:
+        yield UrllibStream(url=url, response=response)
+
+
+class CurlStream(UrlStream):
+    """Streams a file from a URL using curl."""
+
+    def __init__(self, *, url: str, headers: HTTPMessage, effective_url: str, stream):
+        self.url = url
+        self.headers = headers
+        self._effective_url = effective_url
+        self._stream = stream
+
+    def read(self, size: int) -> bytes:
+        return self._stream.read(size)
+
+    def geturl(self) -> str:
+        return self._effective_url
+
+
+def curl_cookie_args(cookie) -> List[str]:
+    """Returns the arguments to pass to curl to use a cookie.
+
+    This can be used as a helper function to generate the arguments for the constructor.
+    """
+    return ["-j", "-b", cookie]
+
+
+@contextlib.contextmanager
+def curl_stream(
+    *, url: str, timeout: int, config_args: Optional[List[str]] = None
+) -> Generator[UrlStream, Any, None]:
+    config_args = config_args or []
+    cmd = [which_string("curl", required=True)]
+    cmd.extend(base_curl_fetch_args(url, timeout, status_bar=False, user_agent=SPACK_USER_AGENT))
+    cmd.extend(config_args)
+    with _start_curl_process(cmd) as curl_process:
+        scheme = urllib.parse.urlparse(url).scheme
+        assert curl_process.stdout is not None, "curl process stdout is None"
+        stream = io.BufferedReader(curl_process.stdout)  # type: ignore[type-var]
+        effective_url = url
+        if scheme not in ("http", "https"):
+            headers = HTTPMessage()
+        else:
+            # curl echoes intermediate redirect responses, so we might get multiple responses
+            finished = False
+            while not finished:
+                headers, finished = _get_next_http_headers(url=url, stream=stream)
+                effective_url = headers.get("location", effective_url)
```

Comment:
I would merge this into `_get_next_http_headers` so you can unit test it as a whole, without having to call `curl_stream`.